### PR TITLE
Shender/environment dashboard

### PR DESCRIPTION
**Pull request by zendesk-shender**
Display versions of projects deployed across our environments.

/cc @zendesk/samson @grosser @sandlerr @pswadi-zendesk @jonmoter 
production
Yellow line denotes something to look at, because there are non-unique versions deployed to these deploy-groups.
master

### References
 - Jira link: https://zendesk.atlassian.net/browse/RUN-11

### Risks
 - None

### CODE REVIEW

File: app/models/project.rb [hunk 1]
```diff
@@ -100,8 +100,22 @@ def with_lock(output: StringIO.new, holder:, error_callback: nil, timeout: 10.mi
     MultiLock.lock(id, holder, timeout: timeout, failed_to_lock: callback, &block)
   end
 
+  def last_release_by_deploy_group
```
Comment (grosser) [inline app/models/project.rb]: has nothing to do with releases -> `last_deploy_by_group`

File: app/controllers/dashboards_controller.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class DashboardsController < ApplicationController
+  def show
+    @environment = Environment.find(params[:id])
+    @data = Project.alphabetical.each_with_object({}) do |project, hash|
+      hash[project] = project.last_release_by_deploy_group
```
Comment (grosser) [inline app/controllers/dashboards_controller.rb]: we could avoid some sql work by passing in the groups we are interested in `last_by_group(@environment.deploy_group_ids)` and then skip the `select!`

File: app/models/project.rb [hunk 3]
```diff
@@ -100,8 +100,22 @@ def with_lock(output: StringIO.new, holder:, error_callback: nil, timeout: 10.mi
...
+    releases = releases_by_deploy_group
+    releases.map { |k, deploys| [ k, deploys.sort_by { |deploy| deploy.updated_at }.reverse.first ] }.to_h
+  end
+
   private
 
+  def releases_by_deploy_group
+    stages.each_with_object({}) do |stage, result|
```
Comment (grosser) [inline app/models/project.rb]: does `include(:deploy_groups)` help lowering the sql ?

File: app/views/dashboards/_project.html.erb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<tr class="<%= 'warning' if deploy_group_versions.values.uniq.count > 1 %>">
```
Comment (grosser) [inline app/views/dashboards/_project.html.erb]: just blow up, something went wrong ?

File: app/views/dashboards/_project.html.erb [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+  <td><%= link_to project.name, project_path(project) %></td>
```
Comment (grosser) [inline app/views/dashboards/_project.html.erb]: `link_to project.name, project` does the same

File: app/views/dashboards/_project.html.erb [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+  <% @environment.deploy_groups.each do |deploy_group| %>
+    <td>
+      <%= link_to deploy_group_versions[deploy_group.id].reference, project_deploy_path(project, deploy_group_versions[deploy_group.id]) if deploy_group_versions[deploy_group.id] %>
```
Comment (grosser) [inline app/views/dashboards/_project.html.erb]: can be a bit dryer using a local variable
```Erb
<% if deploy = deploy_group_versions[deploy_group.id] %>
  ...
<% end %>
```

File: app/views/dashboards/show.html.erb [hunk 7]
```diff
@@ -0,0 +1,20 @@
+<h1><%= @environment.name %></h1>
+
+<section>
+  <table class="table">
```
Comment (grosser) [inline app/views/dashboards/show.html.erb]: that looks suspicious, do we have tables without class table ? :D

File: app/views/dashboards/show.html.erb [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    <tr>
+      <th></th>
+      <% @environment.deploy_groups.each do |deploy_group| %>
+        <th><%= deploy_group.name %></th>
+      <% end %>
+    </tr>
+    </thead>
+    <tbody>
+    <% @data.each do |project, deploy_group_versions| %>
+      <%= render partial: 'project', object: project,
```
Comment (grosser) [inline app/views/dashboards/show.html.erb]: you might be able to get away with `render project, locals: { deploy_group_versions: deploy_group_versions }`

Comment (grosser) [inline app/views/dashboards/show.html.erb]: nah never mind ... that would render `project/_project`

File: app/views/shared/_header.html.erb [hunk 9]
```diff
@@ -5,34 +5,48 @@
 
     <ul class="nav navbar-nav">
       <% if current_user %>
-          <li class="dropdown <%= 'active' if request.path.starts_with?('/projects') %>">
-            <a href="#" class="dropdown-toggle" data-toggle="dropdown">Projects <b class="caret"></b></a>
+        <li class="dropdown <%= 'active' if request.path.starts_with?('/projects') %>">
+          <a href="#" class="dropdown-toggle" data-toggle="dropdown">Projects <b class="caret"></b></a>
+          <ul class="dropdown-menu">
+            <% current_user.starred_projects.alphabetical.each do |project| %>
+                <li>
```
Comment (grosser) [inline app/views/shared/_header.html.erb]: too much indent

File: app/views/shared/_header.html.erb [hunk 10]
```diff
@@ -5,34 +5,48 @@
...
+                  <%= link_to project_path(project) do %>
+                      <span class="glyphicon glyphicon-star"></span> <%= project.name %>
+                  <% end %>
+                </li>
+            <% end %>
+            <% (Project.alphabetical.to_a - current_user.starred_projects).each do |project| %>
+                <li>
+                  <%= link_to project_path(project) do %>
+                      <span class="glyphicon">&nbsp;</span> <%= project.name %>
```
Comment (grosser) [inline app/views/shared/_header.html.erb]: too much indent

File: app/views/shared/_header.html.erb [hunk 11]
```diff
@@ -5,34 +5,48 @@
...
-              <% (Project.alphabetical.to_a - current_user.starred_projects).each do |project| %>
-                  <li>
-                    <%= link_to project_path(project) do %>
-                        <span class="glyphicon">&nbsp;</span> <%= project.name %>
-                    <% end %>
-                  </li>
+              <% Environment.all.each do |env| %>
+                <li>
+                  <%= link_to dashboard_path(env) do %>
+                    <span class="glyphicon">&nbsp;</span> <%= env.name %>
```
Comment (grosser) [inline app/views/shared/_header.html.erb]: this `class="glyphicon"` thing is reused twice, maybe put it into a descriptive helper ?

File: app/views/shared/_header.html.erb [hunk 12]
```diff
@@ -5,34 +5,48 @@
...
-            <% end %>
-          </li>
+        <% end %>
+        <li class="<%= 'active' if current_page?(controller: '/deploys', action: 'recent') %>">
+          <%= link_to "Recent Deploys", recent_deploys_path %>
+        </li>
+        <li class="<%= 'active' if current_page?(controller: '/deploys', action: 'active') %>">
+          <%= link_to active_deploys_path do %>
+              Current Deploys
+              <%= content_tag(:span, Deploy.active.count, class: "badge") unless Deploy.active.count.zero? %>
```
Comment (grosser) [inline app/views/shared/_header.html.erb]: this is getting the count twice ->

```Ruby
count = Deploy.active.count
content_tag(:span, count, class: "badge") unless count.zero?
```
```

File: test/controllers/dashboards_controller_test.rb [hunk 13]
```diff
@@ -0,0 +1,21 @@
+require 'test_helper'
+
+describe DashboardsController do
+  let(:production) { environments(:production_env) }
+
+  def self.it_renders_show
+    it 'get :show succeeds' do
```
Comment (grosser) [inline test/controllers/dashboards_controller_test.rb]: more descriptive + real english would be `it 'renders show'` :)

File: test/models/project_test.rb [hunk 14]
```diff
@@ -238,7 +238,27 @@
       output.string.include?('using the error callback').must_equal(true)
       output.string.include?("Can't get here").must_equal(false)
     end
-
   end
 
+  describe 'deploy group releases' do
```
Comment (grosser) [inline test/models/project_test.rb]: using `describe "#last_release_by_deploy_group"` makes it easier to see what method this is for

File: test/models/project_test.rb [hunk 15]
```diff
@@ -238,7 +238,27 @@
...
+    let(:prod_deploy) { deploys(:succeeded_production_test) }
+    let(:staging_deploy) { deploys(:succeeded_test) }
+
+    it 'contains releases per deploy group' do
+      project.last_release_by_deploy_group[deploy_group_pod1.id].must_equal prod_deploy
+      project.last_release_by_deploy_group[deploy_group_pod2.id].must_equal prod_deploy
+      project.last_release_by_deploy_group[deploy_group_pod100.id].must_equal staging_deploy
+    end
+
+    it 'contains no releases per deploy group' do
```
Comment (grosser) [inline test/models/project_test.rb]: `for undeployed projects`

**Comment by grosser** [comment]
logic looks good, I'm a little afraid of n+1 queries, check if you can reduce the number of queries with a few nice includes, potentially add a test like `assert_sql_queries 10 do` (with query_diet + steal code from classic)

File: app/controllers/dashboards_controller.rb (hunk 2 above)
Comment (zendesk-shender) [inline app/controllers/dashboards_controller.rb]: I debated that, but in the end liked have the last_release_by_deploy_group environment agnostic. I'll change it if you feel strongly about it.

File: app/views/dashboards/_project.html.erb (hunk 4 above)
Comment (zendesk-shender) [inline app/views/dashboards/_project.html.erb]: No, nothing went wrong, it's just a visual cue to the user that there are multiple versions of a project deployed to this environment. It's also a valid state for a slow-rollout deployment strategy.

File: app/views/dashboards/_project.html.erb (hunk 5 above)
Comment (zendesk-shender) [inline app/views/dashboards/_project.html.erb]: ah, thanks!

File: app/views/dashboards/_project.html.erb (hunk 6 above)
Comment (zendesk-shender) [inline app/views/dashboards/_project.html.erb]: Aye, I just can't bring myself to do assignment within a conditional check. Years of c/c++ bugs have rejected that pattern.
I'll have a look at creating a helper method for this.

File: app/views/dashboards/show.html.erb (hunk 7 above)
Comment (zendesk-shender) [inline app/views/dashboards/show.html.erb]: I think this was just to get the bootstrap styling showing correctly. :)

File: app/views/shared/_header.html.erb (hunk 9 above)
Comment (zendesk-shender) [inline app/views/shared/_header.html.erb]: oops.

File: app/views/shared/_header.html.erb (hunk 12 above)
Comment (zendesk-shender) [inline app/views/shared/_header.html.erb]: Ah, this isn't code I changed, so might leave this for another PR.

File: test/models/project_test.rb (hunk 14 above)
Comment (zendesk-shender) [inline test/models/project_test.rb]: Aye, tough battle with keeping it consistent with existing code, but yea, since there is no prevailing pattern, I'll do this.

File: test/models/project_test.rb (hunk 15 above)
Comment (zendesk-shender) [inline test/models/project_test.rb]: done.

File: test/controllers/dashboards_controller_test.rb (hunk 13 above)
Comment (zendesk-shender) [inline test/controllers/dashboards_controller_test.rb]: done.

File: app/models/project.rb (hunk 1 above)
Comment (zendesk-shender) [inline app/models/project.rb]: done.

**Comment by zendesk-shender** [comment]
@grosser , fixed most of the comments, I'm not sure how to fix the stages->deploys n+1 queries as the stages would have to preload a limited scope of deploys. I'm looking into it, and at least learning stuff! :)

And thanks for the pointer to querydiet, I'm using that for testing now.

File: app/views/dashboards/_project.html.erb (hunk 6 above)
Comment (zendesk-shender) [inline app/views/dashboards/_project.html.erb]: Ah well, couldn't find a cleaner way, so yup, implemented!

File: app/controllers/dashboards_controller.rb [hunk 16]
```diff
@@ -0,0 +1,9 @@
+class DashboardsController < ApplicationController
+  def show
+    @environment = Environment.find(params[:id])
+    @data = Project.alphabetical.with_deploy_groups.each_with_object({}) do |project, hash|
```
Comment (grosser) [inline app/controllers/dashboards_controller.rb]: `@data` is a bit vague, maybe `@deploys`

File: app/models/project.rb [hunk 17]
```diff
@@ -100,8 +101,22 @@ def with_lock(output: StringIO.new, holder:, error_callback: nil, timeout: 10.mi
     MultiLock.lock(id, holder, timeout: timeout, failed_to_lock: callback, &block)
   end
 
+  def last_deploy_by_group
+    releases = deploys_by_group
+    releases.map { |k, deploys| [ k, deploys.sort_by { |deploy| deploy.updated_at }.reverse.first ] }.to_h
```
Comment (grosser) [inline app/models/project.rb]: can use proc shorthand for sort and reverse.first == last
`releases.map { |group_id, deploys| [ group_id, deploys.sort_by(&:updated_at).last ] }.to_h`

File: test/models/project_test.rb [hunk 18]
```diff
@@ -238,7 +238,36 @@
       output.string.include?('using the error callback').must_equal(true)
       output.string.include?("Can't get here").must_equal(false)
     end
-
   end
 
+  describe '.last_deploy_by_group' do
```
Comment (grosser) [inline test/models/project_test.rb]: `#last_deploy_by_group` since it's an instance method

File: test/models/project_test.rb [hunk 19]
```diff
@@ -238,7 +238,36 @@
...
+      project = Project.create!(name: 'blank_new_project', repository_url: url)
+
+      project.last_deploy_by_group[deploy_group_pod1.id].must_be_nil
+      project.last_deploy_by_group[deploy_group_pod2.id].must_be_nil
+      project.last_deploy_by_group[deploy_group_pod100.id].must_be_nil
+    end
+
+    it 'performs minimal number of queries' do
+      Project.create!(name: 'blank_new_project', repository_url: url)
+      assert_sql_queries 7 do
```
Comment (grosser) [inline test/models/project_test.rb]: sweet!

File: test/models/project_test.rb [hunk 20]
```diff
@@ -238,7 +238,36 @@
...
+        Project.alphabetical.with_deploy_groups.each_with_object({}) do |p, hash|
+          hash[p] = p.last_deploy_by_group
```
Comment (grosser) [inline test/models/project_test.rb]: not sure why we need a hash assignment, the result is not used ?

File: test/test_helper.rb [hunk 21]
```diff
@@ -28,6 +28,7 @@ class ActiveSupport::TestCase
   before do
     Rails.cache.clear
     stubs_project_callbacks
+    QueryDiet::Logger.reset
```
Comment (grosser) [inline test/test_helper.rb]: we don't need this since we always look at the diff anyway

File: test/test_helper.rb [hunk 22]
```diff
@@ -48,6 +49,28 @@ def unstub_project_callbacks
     Project.any_instance.unstub(:clean_repository)
   end
 
+  def map_ar_queries
+    QueryDiet::Logger.queries.map(&:first) - ["select 1"]
+  end
+
+  def ar_queries
```
Comment (grosser) [inline test/test_helper.rb]: don't need these 1-off methods for now, `ar_queries` can just return the sql array

**Comment by grosser** [comment]
:+1:

File: test/models/project_test.rb (hunk 18 above)
Comment (zendesk-shender) [inline test/models/project_test.rb]: Ah, was wondering what the distinction was. Thanks.

File: app/models/project.rb (hunk 17 above)
Comment (zendesk-shender) [inline app/models/project.rb]: Cool, done.

File: test/models/project_test.rb (hunk 20 above)
Comment (zendesk-shender) [inline test/models/project_test.rb]: Ah, nice catch. Yup, cut/paste mistake.

File: test/test_helper.rb (hunk 21 above)
Comment (zendesk-shender) [inline test/test_helper.rb]: Ah, good point.

File: app/views/shared/_header.html.erb (hunk 11 above)
Comment (zendesk-shender) [inline app/views/shared/_header.html.erb]: Left this for now, as I think the HTML is more readable with the code inline.

**Comment by zendesk-shender** [comment]
merged
